src/rta/fullsupernodelist.cpp

```diff
@@ -226,84 +229,64 @@ SupernodePtr FullSupernodeList::get(const string &address) const
     return SupernodePtr(nullptr);
 }
 
-bool FullSupernodeList::buildAuthSample(uint64_t height, vector<SupernodePtr> &out)
+bool FullSupernodeList::buildAuthSample(uint64_t height, const std::string& payment_id, vector<SupernodePtr> &out)
 {
-    crypto::hash block_hash;
-    string  block_hash_str;
-
-    MDEBUG("building auth sample for height: " << height);
+    boost::shared_lock<boost::shared_mutex> readerLock(m_access);
 
-    if (!getBlockHash(height - AUTH_SAMPLE_HASH_HEIGHT, block_hash_str)) {
-        LOG_ERROR("getBlockHash error");
+    if (height != m_blockchain_based_list_block_number)
+    {
+        LOG_ERROR("unable to build auth sample for block height " << height << " and PaymentID "
+          << payment_id << " because of mistmatch with blockchain based list block number " << m_blockchain_based_list_block_number);
         return false;
     }
 
-    epee::string_tools::hex_to_pod(block_hash_str, block_hash);
+    MDEBUG("building auth sample for height " << height << " and PaymentID " << payment_id);
 
-    std::array<std::vector<SupernodePtr>, TIERS> tier_supernodes;
+    out.clear();
+    out.reserve(AUTH_SAMPLE_SIZE);
+
+    const unsigned char* it               = reinterpret_cast<const unsigned char*>(payment_id.c_str());
+    const size_t         supernodes_count = m_blockchain_based_list.size();
+
+    for (bool loop=true; loop && *it && out.size() < AUTH_SAMPLE_SIZE; it++)
     {
-        boost::shared_lock<boost::shared_mutex> readerLock(m_access);
-        int64_t now = static_cast<int64_t>(std::time(nullptr));
-        int64_t cutoff_time = now - ANNOUNCE_TTL_SECONDS;
-        for (const auto &sn_pair : m_list) {
-            const auto &sn = sn_pair.second;
-            const auto tier = sn->tier();
-            MTRACE("checking supernode " << sn_pair.first << ", updated: " << (now - sn->lastUpdateTime()) << "s ago"
-                   << ", tier: " << tier);
-            if (tier > 0 && sn->lastUpdateTime() >= cutoff_time)
-                tier_supernodes[tier - 1].push_back(sn);
-        }
-    }
+        for (size_t offset=0;; offset++)
+        {    
+            if (offset == supernodes_count)
+            {
+                loop = false;
+                break; //all supernodes have been selected
+            }
+
+            size_t       supernode_index  = (offset + size_t(*it)) % supernodes_count;
```

Comment:
This is highly non-random, and selects from a very small set of the supernodes; given a UUID such as

`xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx` where the `x`s are hex digits, this means this algorithm:

- select a supernode (sort of) randomly from supernodes [48-57, 97-102]
- select a supernode (sort of) randomly from supernodes [49-58, 98-103]
- ...
- select a supernode (sort of) randomly from supernodes [55,57,59,61,63,104,106,108]

with some adjustments in case of conflicts.